clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java

```diff
@@ -1610,7 +1589,7 @@ private void updateLastSeenEpochIfNewer(TopicPartition topicPartition, OffsetAnd
     public boolean updateAssignmentMetadataIfNeeded(Timer timer) {
         maybeThrowFencedInstanceException();
         offsetCommitCallbackInvoker.executeCallbacks();
-        maybeUpdateSubscriptionMetadata();
+        applicationEventHandler.addAndGet(new UpdateSubscriptionMetadataEvent(calculateDeadlineMs(timer)));
```

Comment:
I think the update of the metadata is expected to be a non-blocking operation which takes the latest information it has available. I am not convinced this should be an `addAndGet` as a result.

If this does now start applying the timer to two operations, I would do a `timer.update()` before the `updateFetchPositions`.

Reply:
Hi @AndrewJSchofield, thanks for the review. To align with old behavior, I think we may need to use `addAndGet` here. I add a `timer.update()` before the `updateFetchPositions`.

In `ClassicKafkaConsumer`, it does a blocking operation to update assignment metadata and a non-blocking operation to wait for joining group.

https://github.com/apache/kafka/blob/553e6b4c6d912b3b619b5a1a5263b616b6d3f62e/clients/src/main/java/org/apache/kafka/clients/consumer/internals/ClassicKafkaConsumer.java#L608-L621

In `ApplicationEventProcessor#process(UpdateSubscriptionMetadataEvent)`, we do a similar behavior.

https://github.com/apache/kafka/blob/76a64a8af60e4035219e016014f2f82534267d1a/clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/ApplicationEventProcessor.java#L285-L297